contracts/protocol/modules/CustomOracleNAVIssuanceModule.sol

```diff
@@ -0,0 +1,1126 @@
+        view
+        returns (uint256)
+    {
+        uint256 existingUnit = _setToken.getDefaultPositionRealUnit(_reserveAsset).toUint256();
+        uint256 totalReserve = existingUnit
+            .preciseMul(_issueInfo.previousSetTokenSupply)
+            .add(_issueInfo.netFlowQuantity);
+
+        return totalReserve.preciseDiv(_issueInfo.newSetTokenSupply);
+    }
+
+    /**
+     * The new position reserve asset unit is calculated as follows:
+     * totalReserve = (oldUnit * oldSetTokenSupply) - reserveQuantityToSendOut
+     * newUnit = totalReserve / newSetTokenSupply
+     */ 
+    function _getRedeemPositionUnit(
+        ISetToken _setToken,
+        address _reserveAsset,
+        ActionInfo memory _redeemInfo
+    )
+        internal
+        view
+        returns (uint256)
+    {
+        uint256 existingUnit = _setToken.getDefaultPositionRealUnit(_reserveAsset).toUint256();
+        uint256 totalExistingUnits = existingUnit.preciseMul(_redeemInfo.previousSetTokenSupply);
+
+        uint256 outflow = _redeemInfo.netFlowQuantity.add(_redeemInfo.protocolFees).add(_redeemInfo.managerFee);
+
+        // Require withdrawable quantity is greater than existing collateral
+        require(totalExistingUnits >= outflow, "Must be greater than total available collateral");
+
+        return totalExistingUnits.sub(outflow).preciseDiv(_redeemInfo.newSetTokenSupply);
+    }
+
+    /**
+     * If a pre-issue hook has been configured, call the external-protocol contract. Pre-issue hook logic
+     * can contain arbitrary logic including validations, external function calls, etc.
+     */
+    function _callPreIssueHooks(
+        ISetToken _setToken,
+        address _reserveAsset,
+        uint256 _reserveAssetQuantity,
+        address _caller,
+        address _to
+    )
+        internal
+    {
+        INAVIssuanceHook preIssueHook = navIssuanceSettings[_setToken].managerIssuanceHook;
+        if (address(preIssueHook) != address(0)) {
+            preIssueHook.invokePreIssueHook(_setToken, _reserveAsset, _reserveAssetQuantity, _caller, _to);
+        }
+    }
+
+    /**
+     * If a pre-redeem hook has been configured, call the external-protocol contract.
+     */
+    function _callPreRedeemHooks(ISetToken _setToken, uint256 _setQuantity, address _caller, address _to) internal {
+        INAVIssuanceHook preRedeemHook = navIssuanceSettings[_setToken].managerRedemptionHook;
+        if (address(preRedeemHook) != address(0)) {
+            preRedeemHook.invokePreRedeemHook(_setToken, _setQuantity, _caller, _to);
+        }
+    }
+
+    /**
+     * If a custom set valuer has been configured, use it. Otherwise fetch the default one form the
+     * controller.
+     */
+    function _getSetValuer(ISetToken _setToken) internal view returns (ISetValuer) {
+        ISetValuer customValuer =  navIssuanceSettings[_setToken].setValuer;
+        return address(customValuer) == address(0) ? controller.getSetValuer() : customValuer;
```

Comment:
What's the thinking here? If the oracle is not present this will fall back and fail anyway. You can save some gas by eliminating this altogether and requiring a valuer

Reply:
Initially it was implemented as an improvement to the NAVIssuanceModule, and then @bweick proposed we move it to a different contract to keep track of versions more easily.

What I imagine is this contract being added to the controller and the old one deprecated

> If the oracle is not present this will fall back and fail anyway.

You mean if the setValuer isn't set in the controller? If the set manager chooses to delegate the set valuation to Set, then I think it's fair to assume that Set can then break it.